WIP:  Support Elasticserch v5.6.4

eeeebbbbrrrr opened the pull request:
This pull request, which addresses issue #114, is based on the work from ZDB v4.0.0.  It ports ZDB to ES 5.6.4.

There's still some cleanup work and changes I want to make to better adhere to the new ES 5.6.3 Plugin API, but right now, all junit and postgres regression tests pass.

I'll attach some artifacts in case anyone wants to help test a little bit before I release this.

eeeebbbbrrrr commented:
ugh.  Travis is failing, it seems, because Elasticsearch isn't starting, despite the fact `sudo /etc/init.d/elasticsearch start` indicates it is.  I'll look into this more tomorrow and will upload binary artifacts once they finish building.

eeeebbbbrrrr commented:
Here are the binary artifacts if anyone wants to test.  I would not recommend this for production use yet!

[artifacts.zip](https://github.com/zombodb/zombodb/files/1440272/artifacts.zip)

eeeebbbbrrrr commented:
Here's updated artifacts that include the performance optimizations I've been working on the past few days.  In general, everything is measurably faster.  Postgres regression test times have decreased by a bit over 15s since v3.2.0 that uses ES 1.7.

Still don't recommend this for production use yet -- I still haven't been able to figure out why Travis-CI keeps failing.

[artifacts.zip](https://github.com/zombodb/zombodb/files/1452007/artifacts.zip)

webattitude commented:
elasticsearch is not found, my guess is `before_install` and `script` sections don't use same working directory else it is not `~`... you may use `TRAVIS_BUILD_DIR`

eeeebbbbrrrr commented:
> elasticsearch is not found, my guess is before_install and script sections don't use same working directory or else it is not ~ ($HOME)... you may use $TRAVIS_BUILD_DIR

Thanks for the tip, @webattitude.  I'll dig into it again tomorrow.  Fighting with Travis is my least favorite thing to do in the entire universe.

eeeebbbbrrrr commented:
Here's an overview of the stuff I've been working on with this PR.  Combined, these things have really improved ZomboDB's performance in basically all situations.  I'm pretty excited about this and am planning to get all this released around the 16th of November.

 - Support ES 5.6.4 only
 - Cleaned up JUnit test framework that validates generated Query DSL JSON
 - Removed dead code
 - Expose Elasticsearch's query profile feature via a new function named `zdb_profile_query(table_name::regclass, query::text)` (e18315d)
 - Add a field called `_zdb_id` in every index which is a 64bit encoded version of the Postgres `ctid` system column.  This field is used for `SELECT` statements to quickly decode and return the matching docs (40fadb3)
 - If a `SELECT` statement doesn't use `zdb_score()` then don't bother generating scores.  This saves CPU cycles in ES along with 4 bytes per row across the network for every document found (5b3ee92)
 - Eliminate spurious Postgres query ZDB executed (via SPI) every time it did anything (5b3ee92)
 - Enable keep-alive requests with libcurl and avoid re-initialization of the primary curl object (5b3ee92)
 - Change how ZDB estimates the number of rows that'll be returned by a `SELECT` by defaulting to 2500.  This eliminates a round-trip to Elasticsearch for every `SELECT` that used to essentially execute a `_count`.  There are index and session-level properties that can be set to hardcode a different value or force the old behavior of doing a `_count`.  Unlike Postgres' built-in index types (btree, gin, etc), it'll always be faster for Postgres to plan an Index Scan when a ZomboDB index is involved, rather than a Sequential Scan. (8f0707f)
 - Ensure that we always sort `_zdb_id` values, in Elasticsearch, before returning them to Postgres.  And where we can, do that sorting in parallel across the cluster.  This allows ZomboDB and Postgres to work together such that Postgres walks the heap in a sequential manner rather than a random-access pattern (47f7526, bcc4381)
 - Create a new Elasticsearch shard action called `FastTerms` that can very quickly retrieve `doc_values` for a field from docs that match a query.  This is used, when scores aren't needed, by normal `SELECT` statements and also for solving cross-index joins via index linking (3243aa2)
 - Add an index option called `store` that will tell ZomboDB+Elasticsearch to store the `_source` for every row in the index.  This will allow you to put something like Kibana on top of the indexes ZomboDB creates.

I'm building binary artifacts now and will upload once complete.

eeeebbbbrrrr commented:
[artifacts.zip](https://github.com/zombodb/zombodb/files/1460479/artifacts.zip)

eeeebbbbrrrr commented:
Travis-CI is great software but boy is it nearly impossible to debug problems.  But, it's finally green!

eeeebbbbrrrr commented:
[artifacts.zip](https://github.com/zombodb/zombodb/files/1481225/artifacts.zip)

I'd really appreciate it if someone following this pull request could take a few minutes and test this out.  I've done a significant amount of testing but could use another pair of eyes or two.

webattitude commented:
How can I help ?

eeeebbbbrrrr commented:
> How can I help ?

It'd be cool if you could download the artifacts from my previous comment, install it on a multi-node ES cluster, and throw some data and queries at it.

I feel like it's solid now and ready to be released (once I fix #235 and #233 ), but having someone else do some testing with data I've never seen before would be a great confidence boost.  There's a *ton* of changes here to support ES 5.6 and otherwise optimize things.

If you'd like to chat offline about details, I'm happy to do a Google Hangout or something.

eeeebbbbrrrr commented:
I also saw Postgres crash once during development on my Mac.  That's what prompted 
8a5b2a5 (I undid that somewhere early on in this PR and it was a mistake) and I haven't seen it crash since. 

(short pause as I remembered I never looked at the core dump for that)

it crashed inside libcurl:

```
  * frame #0: 0x00007fff60b7bfce libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff60cb9150 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff60ad832a libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff60be0b28 libsystem_malloc.dylib`szone_error + 596
    frame #4: 0x00007fff60bd676b libsystem_malloc.dylib`tiny_free_list_remove_ptr + 298
    frame #5: 0x00007fff60beb690 libsystem_malloc.dylib`tiny_free_no_lock + 370
    frame #6: 0x00007fff60bec254 libsystem_malloc.dylib`free_tiny + 628
    frame #7: 0x00007fff5f343bd5 libcurl.4.dylib`Curl_llist_remove + 124
    frame #8: 0x00007fff5f357dff libcurl.4.dylib`Curl_conncache_remove_conn + 64
    frame #9: 0x00007fff5f330b85 libcurl.4.dylib`Curl_disconnect + 168
    frame #10: 0x00007fff5f344b18 libcurl.4.dylib`multi_done + 407
    frame #11: 0x00007fff5f346195 libcurl.4.dylib`multi_runsingle + 2734
    frame #12: 0x00007fff5f3455ed libcurl.4.dylib`curl_multi_perform + 114
    frame #13: 0x00007fff5f33e741 libcurl.4.dylib`curl_easy_perform + 362
    frame #14: 0x0000000103943c51 zombodb.so`rest_call + 769
    frame #15: 0x000000010393d7f6 zombodb.so`elasticsearch_estimateCount + 134
    frame #16: 0x0000000103939a7d zombodb.so`wrapper_estimateCount + 93
    frame #17: 0x000000010393f714 zombodb.so`zdb_internal_estimate_count + 68
    frame #18: 0x000000010334fb31 postgres`ExecMakeFunctionResultNoSets(fcache=0x00007fbcfe847490, econtext=<unavailable>, isNull="\x7f~\x7f\x7f\x7f\x7f\x7f\x7f", isDone=<unavailable>) at execQual.c:2027 [opt]
```

hmm.  This one is kinda concerning.  I changed how ZDB uses libcurl such that it only needs to instantiate the libcurl object once per backend session instead of every query.  

I'll spend some time trying to re-create this (although I no good ideas how) and otherwise audit all that code.

eeeebbbbrrrr commented:
So this line is bugged:
   https://github.com/zombodb/zombodb/blob/42e71fac950b3f87ba679d4350b713e41725e3b4/postgres/src/main/c/rest/rest.c#L305

Since I'm using one global curl instance, that needs to be allocated with malloc (well, calloc), so I'm going to push a fix for that right now, but I don't know if that would cause the segfault I saw.  

I've asked on the curl-library mailing list if they might have any ideas about that backtrace.

eeeebbbbrrrr commented:
Commit 672af81 implemented a new Lucene Query that I called "CrossJoin".  It's akin to SIREn's "filter_join" query, but a clean-room implementation.  And many subsequent commits worked on optimizing it.  Combined with the "FastTerms" action I implemented prior, it's what ZDB now uses when doing `#expand()` clauses and for queries that use ZDB's "index linking" feature.

Looks like it's slightly slower compared to the SIREn plugin on ES 1.7, so I'm going to spend a little more time optimizing things.  Something around 700ms for SIREn to join 5M rows on a "long" key v/s about 1000ms for CrossJoin.

I've got some ideas when joining on integer and long fields that I think when done will put "CrossJoin" a good 3x-4x faster than SIREn.  It's just going to take a few days to get all that implemented.

eeeebbbbrrrr commented:
77c1e21 implements what I was talking about above.  It's maybe roughly 2.5x faster than SIREn on ES 1.7.  I think there's room for more optimization tho and am going work on that tomorrow.

eeeebbbbrrrr commented:
[artifacts.zip](https://github.com/zombodb/zombodb/files/1485338/artifacts.zip)

Here are updated artifacts that are current to this point.

webattitude commented:
It's a bit complicated for me right now, but I'll check this out ASAP

eeeebbbbrrrr commented:
Updated artifacts

[artifacts.zip](https://github.com/zombodb/zombodb/files/1489755/artifacts.zip)

eeeebbbbrrrr commented:
I think I'm done.  Everything is at a point where I'd be happy to release it.

That said, I have another idea on how to improve cross-index joining performance even more by taking a different approach.  If it works out, it'll require a REINDEX, so I don't want to release what I've got here until I know if my idea will work or not.

I'm going to start on it tomorrow and then take the rest of the week off for Thanksgiving.  So next week I'll know for certain if it'll gunna pan out the way I want.  ZomboDB's cross-index joining feature is really powerful and after all the work that's gone into general performance against ES 5.6 I want to make this specific feature as best as I can.

eeeebbbbrrrr commented:
Yeah, my ideas for improving cross-index join performance (at least for 32bit/64bit integers) is working out really well.  My little 5-node cluster at home can now join 5M rows in about 110 milliseconds, depending on how randomly distributed the primary key values are.

I still have more work to do to finish this up, but I think I can get it done and everything released by the end of this week.

eeeebbbbrrrr commented:
Here are updated artifacts.  At this point I'm ready to release, but need to update documentation and write release notes.  Be on the look out for a release by the end of this week.

[artifacts.zip](https://github.com/zombodb/zombodb/files/1528554/artifacts.zip)

eeeebbbbrrrr commented:
I think the last change I'm going to make to this is with the version number.  During development of this PR I've had it set to `5.6.4-0`, where `5.6.4` represents the Elasticsearch version it supports, and `-0` is the ZDB version.

I'm going to change this to `5.6.4-1.0.0` where `-1.0.0` represents the ZDB version relative to the ES version.  This way as elastic.co releases more point releases and I release point releases, it'll get really confusing really quickly!  :)

eeeebbbbrrrr commented:
found a bug, need to investigate later this week:

```sql
joins=# vacuum main;
ERROR:  rc=400; No handler found for uri [//_cat/indices/joins.public.main.idxmain?h=docs.deleted] and method [GET]
Time: 7595.799 ms
```

eeeebbbbrrrr commented:
Also, I want to rename the new `optimize_for_joins` index option to simply 'routing' -- and I need to document this thing anyways.

eeeebbbbrrrr commented:
Soon as Travis-CI finishes its current run, I'm going to get this monster merged and released as a _beta1.
